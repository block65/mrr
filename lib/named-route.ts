import type { Params, ExtractRouteParams } from './types.js';
import { nullOrigin, urlObjectAssign, urlRhs } from './util.js';

type QueryParams = Record<string, string>;

export interface NamedRoute<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => string;
}

export function interpolate<P extends string>(
  path: P,
  params: ExtractRouteParams<P> | Params = {},
): string {
  return path.replace(/\/:(\w+)[?+*]?/g, (_match, token: keyof typeof params) =>
    params[token] ? `/${params[token]}` : '',
  );
}

export function namedRoute<T extends string, Q extends QueryParams = never>(
  pathname: T,
): NamedRoute<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      const newUrl = urlObjectAssign(new URL(origin || nullOrigin), {
        pathname: interpolatedPathname,
        search: searchParams.toString(),
        ...rest,
      });

      return newUrl.origin === nullOrigin.origin
        ? urlRhs(newUrl)
        : newUrl.toString();
    },
  };
}
