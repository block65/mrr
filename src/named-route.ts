export { namedRoute } from '../lib/named-route.js';

export type {
  NamedRoute,
  NamedRouteWithParams,
  NamedRouteWithoutParams,
} from '../lib/named-route.js';
