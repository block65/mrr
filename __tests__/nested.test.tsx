import '@testing-library/jest-dom';
import { render, screen, waitFor } from '@testing-library/react';
import type { FC } from 'react';
import { namedRoute } from '../lib/named-route.js';
import { Route, Router, Routes } from '../src/index.js';
import { LocationDisplay } from './index.test.js';

test('wildcard routes + nested', async () => {
  const userRoot = namedRoute('/users');
  const userView = namedRoute('/users/blah/:userId');

  const ComponentWithUserId: FC<{ userId: string }> = ({
    params: { userId },
  }) => <>userId = {userId}</>;

  const ParamlessComponent: FC = () => <>I am a Paramless Component</>;

  const { asFragment } = render(
    <Router pathname="/users/blah/test1">
      <LocationDisplay />
      <Routes>
        <Route wildcard path={userRoot.path}>
          <h1>inside userRoot</h1>
          <Routes>
            <Route path={userRoot.path}>this should not display</Route>
            <Route
              wildcard
              path={userView.path}
              component={(params) => (
                <h1 data-testid="users">You are user {params.userId}</h1>
              )}
            ></Route>
            <Route path={userView.path} component={ComponentWithUserId} />
            <Route>
              <ParamlessComponent />
            </Route>
            <Route>
              <h1>fail</h1>
            </Route>
          </Routes>
        </Route>

        <Route>
          <h1>fail</h1>
        </Route>
      </Routes>
    </Router>,
  );

  await waitFor(() => screen.getByTestId('users'));

  expect(asFragment()).toMatchSnapshot();
});

test('default route + wildcard routes + nested', async () => {
  const userView = namedRoute('/users/blah/:userId');

  const { asFragment } = render(
    <Router pathname="/users/blah/test1">
      <LocationDisplay />
      <Routes>
        <Route>
          <Routes>
            <Route>
              <Routes>
                <Route
                  wildcard
                  path={userView.path}
                  component={(params) => (
                    <h1 data-testid="users">You are user {params.userId}</h1>
                  )}
                />
                <Route>
                  <h1>fail</h1>
                </Route>
              </Routes>
            </Route>
            <Route>
              <h1>fail</h1>
            </Route>
          </Routes>
        </Route>
      </Routes>
    </Router>,
  );

  await waitFor(() => screen.getByTestId('users'));

  expect(asFragment()).toMatchSnapshot();
});
