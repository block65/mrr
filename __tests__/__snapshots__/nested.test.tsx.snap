// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
<DocumentFragment>
  <div
    data-testid="location-display"
  >
    http://localhost:3000/login
  </div>
  <h1
    data-testid="login"
  >
    You are at the login page
  </h1>
</DocumentFragment>
`;

exports[`default route + wildcard routes + nested 1`] = `
<DocumentFragment>
  <div
    data-testid="location-display"
  >
    http://localhost:3000/users/blah/test1
  </div>
  <h1
    data-testid="users"
  >
    You are user test1
  </h1>
</DocumentFragment>
`;

exports[`wildcard routes + nested 1`] = `
<DocumentFragment>
  <div
    data-testid="location-display"
  >
    http://localhost:3000/users/blah/test1
  </div>
  <h1>
    inside userRoot
  </h1>
  <h1
    data-testid="users"
  >
    You are user test1
  </h1>
</DocumentFragment>
`;
